Add typing support to `get_solo`

Viicos opened the pull request:
As `get_solo` hits the cache, type checkers can't infer the return type of the method correctly. User could use `typing.cast` or similar but having it fixed in the library directly sounds easier. `typing.TypeVar` is used as a workaround until `typing.Self` can be used without making use of `typing_extensions`.

Viicos commented:
If you'd like to, we could add basic typing to the other parts of the library as well

johnthagen commented:
I could see value in adding typing to this library, and enabling mypy checking of the project as a new CI check.

An example you could look at is:

- https://github.com/johnthagen/cppcheck-junit/blob/master/tox.ini#L32

Viicos commented:
Great to hear you want to head into this direction. I added types to the whole library, and also updated some old compatibility code. Let me know if any changes are needed.

johnthagen commented on solo/admin.py:
```diff
@@ -1,25 +1,30 @@
 
 
-class SingletonModelAdmin(admin.ModelAdmin):
+class SingletonModelAdmin(admin.ModelAdmin):  # type: ignore[type-arg]
```
Would using [`django-stubs`](https://github.com/typeddjango/django-stubs) help avoid any of these `# type: ignore`'s?

We'd only want `django-stubs` to be a dev dependency though, not a full one. Perhaps only installed via tox (probably need to pin it in tox config or using something like [`pip-tools`](https://github.com/jazzband/pip-tools).

johnthagen replied:
Oh, I did see you added `django-stubs` later in the PR. Any idea why these and other Django related things seem to still need `# type: ignore`? If there is something missing from `django-stubs`, could potentially open an issue over there. That team is quite responsive.

Viicos replied:
So basically `django-stubs` is making use of generics when possible, and `ModelAdmin` is an example. The main drawback is it is only defined as being generic in the stubs, not at runtime. A placeholder `__class_getitem__` was added upstream for some classes (I think `Manager` and some fields), but `ModelAdmin` is still lacking.

The workaround for application developers is to use [`django-stubs-ext`](https://github.com/typeddjango/django-stubs/tree/master/ext) that will monkeypatch Django at runtime and add `__class_getitem__` definitions when required. However, this does not apply to library development.

I think it could be nice to go with the following:

```python
from django.contrib.admin import ModelAdmin as _ModelAdmin

if TYPE_CHECKING:
    ModelAdmin = _ModelAdmin
else:
    class ModelAdmin(_ModelAdmin):
        def __class_getitem__(cls, *args, **kwargs):
            return cls

_ModelT = TypeVar("_ModelT", bound=SingletonModel)

class SingletonModelAdmin(ModelAdmin[_ModelT]):
    ...
```

But this is quite verbose.

Viicos replied:
With the current solution, users will not be able to do the following:

```python
class MySingletonAdmin(SingletonModelAdmin[MySingleton]):
    ...
```

However, this is a niche feature, I don't see a lot of people making use of it (see this [search result](https://github.com/search?q=language%3Apython+%22ModelAdmin%5B%22&type=code)). Let's keep it as it is for now and if people complain about it the workaround above could still be added

johnthagen commented on solo/models.py:
```diff
@@ -1,63 +1,65 @@
 
 DEFAULT_SINGLETON_INSTANCE_ID = 1
 
+Self = TypeVar("Self", bound="SingletonModel")
```
I'm not 100% clear how this is being used, but could we instead use the official `Self` type that is included in Python 3.11 and backported to older versions using [`typing-extenstions`](https://typing-extensions.readthedocs.io/en/latest/#Self)?

This would add `typing-extentions` as a dependency to the project, but it is a very common dependency. We could add a `TODO` somewhere that says to remove the dependency after dropping support for > 3.11.

Viicos replied:
The current solution I provided was the way of doing this before `Self`. `typing.Self` is a convenience feature, but indeed does add an extra dependency.

Quoting my OP:

> `typing.TypeVar` is used as a workaround until `typing.Self` can be used without making use of `typing_extensions`.

So if it's fine to have `typing-extensions`, I'll go for it!

Viicos replied:
> We could add a `TODO` somewhere that says to remove the dependency after dropping support for > 3.11.

The added Python version check is in my opinion explicit enough. If you want me to add an extra TODO, let me know!

johnthagen commented on solo/models.py:
```diff
@@ -1,63 +1,65 @@
-            return obj
-        cache = get_cache(cache_name)
+            obj, _ = cls.objects.get_or_create(pk=cls.singleton_instance_id)
+            return obj  # type: ignore[return-value]
```
Another example where perhaps `django-stubs` can help us ignore some `# type: ignore`s

johnthagen commented on tox.ini:
```diff
@@ -33,3 +34,11 @@ deps =
+skip_install = true
+deps =
+    mypy
+    django-stubs
```
We probably need to pin these (both of these packages change a lot), either manually here, or using `pip-tools` which would be a little more complex, but is more future proof.

Viicos replied:
Yes especially the stubs package!

johnthagen commented:
> Great to hear you want to head into this direction. I added types to the whole library, and also updated some old compatibility code. Let me know if any changes are needed. As another PR, I think it would be great to have:
> 
> * black + isort/ruff
> * Use PEP 517/518 for packaging
> 
> I you agree with this, I can make separate PRs

Let's make separate PRs for these things.

I'd prefer using Ruff for everything on the first bullet: formatting, linting and isort rules. You can see my example repo here for how I usually configure Ruff:

- https://github.com/johnthagen/python-blueprint?tab=readme-ov-file#code-style-checking
- https://github.com/johnthagen/python-blueprint/blob/main/pyproject.toml#L77

We may need to modify some of these rules for `django-solo` (e.g. we should keep the first diff a bit minimal rather than enabling too many of those rules).

Viicos commented on solo/models.py:
```diff
@@ -1,63 +1,70 @@
+    def delete(self, *args: Any, **kwargs: Any) -> tuple[int, dict[str, int]]:
         self.clear_cache()
-        super(SingletonModel, self).delete(*args, **kwargs)
+        return super().delete(*args, **kwargs)
```
This is a runtime change, but imo better this way

johnthagen replied:
This is not a breaking change though, as previous users simply would not have used a return value at all?

Viicos replied:
Yes, just wanted to make sure you were aware of it :)

Viicos commented on solo/models.py:
```diff
@@ -1,63 +1,70 @@
+            obj, _ = cls.objects.get_or_create(pk=cls.singleton_instance_id)
             obj.set_to_cache()
-        return obj
+        return obj  # type: ignore[return-value]
```
Regarding those ignore, I believe it is an issue with `mypy` (`pyright` understands it correctly).

See this [mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=2883f5233e1a9caeeb949739ee7ee1bd) to understand why it is failing (and [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.328&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDKApgDbAA0UAwiQIYDO9AarSJQOJEpEhIDGlACqIiLNoVLAAUFICyUALxRhCUawAUAIlmaAlDL51GUWbRS00PdZ268%2BAbVkBdXQC4pUT1AAmRYFEsYdXpJXSgAWgA%2BE3cvOKgAOiSDI3oTMF8SWK8wACMAKyI%2BGHpXalSxRzMLHntiMicnGSlDBjSCVDQSIhgwFHVZDNI3Dy8AAVbGCB6ACwzRz19-QIB9ejASMHVDejCoiTJs%2BKg8-MUoHYTTopKEwPV9Y88QIgA3IloSFfhVdVPHp4vGAAVxAKBOBSkQA) where it works).

There is already an open issue regarding this: https://github.com/python/mypy/issues/16547

johnthagen commented on MANIFEST.in:
```diff
@@ -1,5 +1,6 @@
 include *.md
 include LICENSE
 include CHANGES
+include solo/py.typed
```
Could you also update `CHANGES`?

johnthagen commented:
I presume you are using `django-solo` in a production application. Could you test your application with this PR to smoke test that everything here works well for you at runtime?

Viicos commented:
> I presume you are using `django-solo` in a production application. Could you test your application with this PR to smoke test that everything here works well for you at runtime?

We are making use of `django-solo` extensively on an open source project, so I can try it there, I'll let you know tomorrow how it goes

Viicos commented:
Here is a pull request making use of my branch:
- https://github.com/open-formulieren/open-forms/pull/3776

I had to add back `get_cache` as we were making use of it in a 3rd party library (I've also added a deprecation warning).

Regarding mypy, we don't run it yet, but I can already see from my IDE (which uses `pyright`) that `.get_solo` is correctly typed.

johnthagen commented:
Could you add a changelog entry for deprecating `get_cache`?

Viicos commented:
Looking at the remaining type issue, might be another `mypy` bug..

Viicos commented:
This is a mypy bug (works with pyright), but I can't produce a MRE: [playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=c55a3b5934dca1f4d2022075f76d2432). Maybe this only happens when split into multiple files

johnthagen commented:
Thanks for the contribution, @Viicos!
